R/conditions.R

```diff
@@ -79,23 +87,53 @@ stop_incompatible_op <- function(op, x, y, details = NULL, ..., message = NULL,
 }
 
 #' @rdname vctrs-conditions
+#' @param to Type to cast to.
+#' @param result The result of a lossy cast.
 #' @export
-warn_lossy_cast <- function(x, y, locations = NULL, details = NULL, ..., message = NULL, .subclass = NULL) {
-
+stop_lossy_cast <- function(x, to, result,
+                            locations = NULL,
+                            details = NULL,
+                            ...,
+                            message = NULL,
+                            .subclass = NULL) {
   message <- message %||% glue_lines(
-    "Lossy cast from <{vec_ptype_full(x)}> to <{vec_ptype_full(y)}>.",
+    "Lossy cast from <{vec_ptype_full(x)}> to <{vec_ptype_full(to)}>.",
     inline_list("Locations: ", locations),
     details
   )
 
-  warn(
-    message,
-    x = x,
-    y = y,
-    locations = locations,
-    details = details,
-    ...,
-    .subclass = c(.subclass, "vctrs_warning_cast_lossy"),
+  withRestarts(
+    vctrs_restart_error_cast_lossy = function() result,
+    abort(
+      message,
+      x = x,
+      y = to,
+      to = to,
+      result = result,
+      locations = locations,
+      details = details,
+      ...,
+      .subclass = c(.subclass, "vctrs_error_cast_lossy")
+    )
+  )
+}
+#' @rdname vctrs-conditions
+#' @param x_ptype,to_ptype Suppress only the casting errors where `x`
+#'   or `to` match these [prototypes][vec_type].
+#' @export
+suppress_errors_lossy_cast <- function(expr, x_ptype = NULL, to_ptype = NULL) {
```

Comment:
My only question is the name of this function — would it be better as `allow_lossy_cast()`?

Reply:
I like it.